src/rust/kafka/src/lib.rs

```diff
@@ -353,17 +357,50 @@ impl BytesConsumer {
             return Err(ConfigurationError::SubscriptionFailed(e));
         }
 
-        Ok(Self { consumer })
+        Ok(Self {
+            consumer,
+            delay_ms: config.delay_ms,
+        })
     }
 
     #[tracing::instrument(skip(self))]
     pub fn stream(&self) -> impl Stream<Item = Result<Bytes, ConsumerError>> + '_ {
         self.consumer.stream().then(move |res| async move {
-            res.map_err(ConsumerError::from).and_then(|msg| {
-                let mut buf = BytesMut::with_capacity(msg.payload_len());
-                buf.extend_from_slice(msg.payload().ok_or(ConsumerError::PayloadAbsent)?);
-                Ok(buf.freeze())
-            })
+            match res {
+                Ok(msg) => {
+                    let timestamp = msg.timestamp();
+                    if let Some(millis_i64) = timestamp.to_millis() {
+                        let millis = if millis_i64 < 0 {
+                            0
+                        } else {
+                            millis_i64.unsigned_abs()
+                        };
+
+                        let message_ts = SystemTime::UNIX_EPOCH + Duration::from_millis(millis);
+                        let target = message_ts + Duration::from_millis(self.delay_ms);
+
+                        if let Err(e) = target.elapsed() {
+                            let duration = e.duration();
+
+                            tracing::debug!(
```

Comment:
should this be a warn? I assume 'elapsed' only fails when the system time goes backwards?

Reply:
The `Err` branch is load-bearing here, we don't care about the `Ok`. `Ok` means that the current time is greater than the target, `Err` means the current time is less than the target (and therefore we need to wait for `duration`).